File: Data Structures & Algorithms/three-integer-sum/submission-2.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res =new ArrayList<>();

        int n = nums.length;

        for(int i=0;i<n;i++){
            if(i>0 && nums[i]==nums[i-1]){
                continue;
            }

            int l=i+1;
            int r=n-1;

            while(l<r){
                int tsum=nums[l]+nums[r]+nums[i];
                if(tsum>0){
                    r--;
                }else if(tsum<0){
                    l++;
                }else{
                    res.add(Arrays.asList(nums[i],nums[l],nums[r]));
                    l++;
                    r--;
                    while (nums[l] ==nums[l-1] && l<r){
                        l++;
                    }
                }
            }
        }
        
            return res;
    }
}

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-2.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:

        pref=[0]*len(nums)
        pref[0]=1

        suf=[0]*len(nums)
        suf[len(nums)-1]=1
        res=[0]*len(nums)
        
        for i in range(1,len(nums)):
            
            pref[i] =nums[i-1] * pref[i-1]
        
        suf[len(nums)-1]=1

        for i in range(len(nums)-2,-1,-1):
            suf[i] = nums[i+1] * suf[i+1]

        for i in range(len(nums)):
            res[i] = pref[i]* suf[i]
        return res
    
        


        